instrumentation/opentelemetry/batchspanprocessor/logger.go

```diff
@@ -2,35 +2,48 @@ package batchspanprocessor // import "github.com/hypertrace/goagent/instrumentat
 
 // Adapted from go.opentelemetry.io/otel/internal/global#internal_logging.go
 import (
-	"log"
-	"os"
+	"sync"
 
 	"github.com/go-logr/logr"
-	"github.com/go-logr/stdr"
+	"github.com/go-logr/zapr"
+	"go.uber.org/zap"
 )
 
 // The logger uses stdr which is backed by the standard `log.Logger`
 // interface. This logger will only show messages at the Error Level.
-var logger logr.Logger = stdr.New(log.New(os.Stderr, "", log.LstdFlags|log.Lshortfile))
+var (
+	logger logr.Logger
+	once   sync.Once
+)
 
 // Info prints messages about the general state of the API or SDK.
 // This should usually be less then 5 messages a minute.
 func Info(msg string, keysAndValues ...interface{}) {
-	logger.V(4).Info(msg, keysAndValues...)
+	getLogger().V(4).Info(msg, keysAndValues...)
 }
 
 // Error prints messages about exceptional states of the API or SDK.
 func Error(err error, msg string, keysAndValues ...interface{}) {
-	logger.Error(err, msg, keysAndValues...)
+	getLogger().Error(err, msg, keysAndValues...)
 }
 
 // Debug prints messages about all internal changes in the API or SDK.
 func Debug(msg string, keysAndValues ...interface{}) {
-	logger.V(8).Info(msg, keysAndValues...)
+	getLogger().V(8).Info(msg, keysAndValues...)
 }
 
 // Warn prints messages about warnings in the API or SDK.
 // Not an error but is likely more important than an informational event.
 func Warn(msg string, keysAndValues ...interface{}) {
-	logger.V(1).Info(msg, keysAndValues...)
+	getLogger().V(1).Info(msg, keysAndValues...)
+}
+
+// since global variables init happens on package imports,
+// the zap global logger might not be set by then
+// hence doing a lazy initialization
+func getLogger() logr.Logger {
```

Comment:
Should we instead pass logger here https://github.com/hypertrace/goagent/blob/94709dab7c096106f9a8a6777b9400299657f4b2/instrumentation/opentelemetry/init.go#L309 and store that in batch_span process struct? What I am unsure is what this global logger is? Secondly if intend to pass a sampling logger from the agent, how would that work? Does it apply rotation that's configured?

Reply:
This works with whatever logger we pass as of now. I've checked with file logger on TPA. Because we set the global logger here: https://github.com/hypertrace/goagent/blob/main/instrumentation/opentelemetry/init.go#L288. Only case this wont work is when someone sets another zap logger from another go routine. But I dont think we do that anywhere. I didnt want to set it in the bsp struct because that'll modify the struct compared to otel and thought we didnt want to do that. @tim-mwangi ?

Reply:
traceable goagent has it's own global logger implemented with zap, I thought that when we port this to that repo, we can change the bsp to our global logger instead of zap global logger and this as makeshift mechanism for the time being